clang-tools-extra/clangd/unittests/PrerequisiteModulesTest.cpp

```diff
@@ -0,0 +1,355 @@
+
+    return Inputs;
+  }
+
+  SmallString<256> TestDir;
+  // Noticed MockFS but its member variable 'OverlayRealFileSystemForModules'
+  // implies that it will better to use RealThreadsafeFS directly.
+  RealThreadsafeFS TFS;
+
+  DiagnosticConsumer DiagConsumer;
+};
+
+TEST_F(PrerequisiteModulesTests, PrerequisiteModulesTest) {
+  MockDirectoryCompilationDatabase CDB(TestDir, TFS);
+
+  CDB.addFile("foo.h", R"cpp(
+inline void foo() {}
+  )cpp");
+
+  CDB.addFile("M.cppm", R"cpp(
+module;
+#include "foo.h"
+export module M;
+  )cpp");
+
+  CDB.addFile("N.cppm", R"cpp(
+export module N;
+import :Part;
+import M;
+  )cpp");
+
+  CDB.addFile("N-part.cppm", R"cpp(
+// Different name with filename intentionally.
+export module N:Part;
+  )cpp");
+
+  CDB.addFile("bar.h", R"cpp(
+inline void bar() {}
+  )cpp");
+
+  CDB.addFile("L.cppm", R"cpp(
+module;
+#include "bar.h"
+export module L;
+  )cpp");
+
+  CDB.addFile("NonModular.cpp", R"cpp(
+#include "bar.h"
+#include "foo.h"
+void use() {
+  foo();
+  bar();
+}
+  )cpp");
+
+  ModulesBuilder Builder(CDB);
+
+  // NonModular.cpp is not related to modules. So nothing should be built.
+  {
+    auto NonModularInfo =
+        Builder.buildPrerequisiteModulesFor(getFullPath("NonModular.cpp"), TFS);
+    EXPECT_TRUE(NonModularInfo);
+    auto Invocation =
+        buildCompilerInvocation(getInputs("NonModular.cpp", CDB), DiagConsumer);
+    EXPECT_TRUE(NonModularInfo->canReuse(*Invocation, TFS.view(TestDir)));
+  }
+
+  {
+    auto MInfo =
+        Builder.buildPrerequisiteModulesFor(getFullPath("M.cppm"), TFS);
+    // buildPrerequisiteModulesFor won't built the module itself.
+    EXPECT_TRUE(MInfo);
+    auto Invocation =
+        buildCompilerInvocation(getInputs("M.cppm", CDB), DiagConsumer);
+    EXPECT_TRUE(MInfo->canReuse(*Invocation, TFS.view(TestDir)));
+  }
+
+  // Module N shouldn't be able to be built.
+  auto NInfo = Builder.buildPrerequisiteModulesFor(getFullPath("N.cppm"), TFS);
+  EXPECT_TRUE(NInfo);
+
+  ParseInputs NInput = getInputs("N.cppm", CDB);
+  std::unique_ptr<CompilerInvocation> Invocation =
+      buildCompilerInvocation(NInput, DiagConsumer);
+  // Test that `PrerequisiteModules::canReuse` works basically.
+  EXPECT_TRUE(NInfo->canReuse(*Invocation, TFS.view(TestDir)));
+
+  // Test that we can still reuse the NInfo after we touch a unrelated file.
+  {
+    CDB.addFile("L.cppm", R"cpp(
+module;
+#include "bar.h"
+export module L;
+export int ll = 43;
+  )cpp");
+    EXPECT_TRUE(NInfo->canReuse(*Invocation, TFS.view(TestDir)));
+
+    CDB.addFile("bar.h", R"cpp(
+inline void bar() {}
+inline void bar(int) {}
+  )cpp");
+    EXPECT_TRUE(NInfo->canReuse(*Invocation, TFS.view(TestDir)));
+  }
+
+  // Test that we can't reuse the NInfo after we touch a related file.
+  {
+    CDB.addFile("M.cppm", R"cpp(
+module;
+#include "foo.h"
+export module M;
+export int mm = 44;
+  )cpp");
+    EXPECT_FALSE(NInfo && NInfo->canReuse(*Invocation, TFS.view(TestDir)));
```

Comment:
can you drop `NInfo &&` part here and below? it looks confusing (we already know its non-null from above, but it won't be obvious when looking at a failure)

Reply:
Done